File: code/jvm/src/main/kotlin/pt/isel/leic/daw/gomokuRoyale/http/controllers/games/models/GamePlayInputModel.kt
package pt.isel.leic.daw.gomokuRoyale.http.controllers.games.models

import pt.isel.leic.daw.gomokuRoyale.domain.Opening
import pt.isel.leic.daw.gomokuRoyale.domain.Piece.Companion.toPiece
import pt.isel.leic.daw.gomokuRoyale.domain.Position
import pt.isel.leic.daw.gomokuRoyale.services.game.GameAction

/**
 * Game move input information
 *
 * @property actionType the type of action being performed
 * @property positionX the x coordinate of the position being played
 * @property positionY the y coordinate of the position being played
 * @property moveChoice the move being chosen
 * @property color the color being chosen
 */

data class GamePlayInputModel(
    val actionType: String, // "PlacePiece", "ChooseMove", or "ChooseColor"
    val positionX: Int?, // used for PlacePiece_x
    val positionY: Int?, // used for PlacePiece_y
    val moveChoice: String?, // used for ChooseMove
    val color: String? // used for ChooseColor
)

fun convertInputModelToGameAction(input: GamePlayInputModel): GameAction {
    return when (input.actionType) {
        "PlacePiece" -> {
            val position = Position(
                input.positionX ?: throw Exception("Invalid position"),
                input.positionY ?: throw Exception("Invalid position")
            )
            GameAction.PlacePiece(position)
        }

        "ChooseMove" -> {
            val move = input.moveChoice ?: throw Exception("Invalid move choice")
            GameAction.ChooseMove(Opening.valueOf(move))
        }

        "ChooseColor" -> {
            input.color?.let { GameAction.ChooseColor(it.toPiece()) } ?: throw Exception("Invalid color")
        }

        else ->
            throw IllegalStateException("Invalid action type")
    }
}

File: code/js/gomoku-app/src/services/game/modals/GamePlayInputModel.ts
export interface GamePlayInputModel {
    actionType: ActionType; // "PlacePiece", "ChooseMove", or "ChooseColor"
    positionX: number | null; // used for PlacePiece_x
    positionY: number | null; // used for PlacePiece_y
    moveChoice: string | null; // used for ChooseMove
    color: string | null; // used for ChooseColor
}

export class GamePlayInputModel {
    constructor(actionType: ActionType, positionX: number, positionY: number, moveChoice: string | null, color: string | null) {
        this.actionType = actionType;
        this.positionX = positionX;
        this.positionY = positionY;
        this.moveChoice = moveChoice;
        this.color = color;
    }
}

export enum ActionType {
    PlacePiece = 'PlacePiece',
    ChooseMove = 'ChooseMove',
    ChooseColor = 'ChooseColor'
}